services/controllerhost/replicaStatsMRUCache.go

```diff
@@ -127,82 +132,134 @@ func (c *replicaStatsMRUCache) get(extent extentID, store storeID) (val *shared.
 		if len(foundStore) != 0 {
 			lclLg.WithField(`substituteStore`, common.FmtStor(string(foundStore))).
 				Error(`Could not get stats for specified store; substituting`)
-			return c.get(extent, foundStore)
+			return psmap.m[foundStore] // Might be nil
 		}
 
 		lclLg.Error(`Could not get stats for specified store`)
 		return nil
-	} else if val != nil {
-		c.put(extent, store, val)
+	} else if extStats != nil {
+		val = makeCompactExtentStats(extStats)
+		// In the typical case, all store extents had planned sealing, meaning that their available sequences are the same
+		// Map the value pointer to the same compactExtentStats in this typical case, to reduce memory usage by nearly 66%
+	dedup:
+		for dedupStore, val2 = range psmap.m {
+			if store != dedupStore && val2 != nil && val2.seqAvailable == val.seqAvailable {
+				val = val2
+				break dedup
+			}
+		}
+
+		// psmap is a pointer into the cache, so we can simply modify it here
+		val.consumerGroupVisibility = psmap.singleCGVisibility
+		psmap.m[store] = val
 	}
 
 	return
-}
 
-func (c *replicaStatsMRUCache) putSingleCGVisibility(extent extentID, singleCGVisibility string) {
-	if len(extent) == 0 || len(singleCGVisibility) == 0 {
-		return
+ttlCheck:
+
+	// Open extents should only be valid for a single scan; All types of sealed extent should have indefinite retention
+	if val != nil && val.status == shared.ExtentStatus_OPEN && time.Since(val.cacheTime) > IntervalBtwnScans {
+		// All the store exents will have the roughly the same cache time, so just delete all of them,
+		// but preserve the singleCGVisibility and the list of storeIDs
+		for s, _ := range psmap.m {
+			psmap.m[s] = nil
+		}
+
+		return c.get(extent, store)
 	}
 
-	c.cache[extent] = newPerStoreReplicaStatsMap(&singleCGVisibility, nil, nil)
+	return
 }
 
-func (c *replicaStatsMRUCache) put(extent extentID, store storeID, val *shared.ExtentStats) {
-	var ok bool
-	var psmap *perStoreReplicaStatsMap
-
-	if len(extent) == 0 || len(store) == 0 {
-		c.log.
-			WithField(common.TagExt, string(extent)). /* Don't call common.Fmt with an empty string ! */
-			WithField(common.TagStor, string(store)).
-			Warn(`Queue depth skipped empty ID`)
+func (c *replicaStatsMRUCache) putPlaceholders(extent extentID, singleCGVisibility string, storeIDs []string) {
```

Comment:
added a comment

`// putPlaceholders is used by processDestination to pre-populate/update some data from the destination extent stats on every pass
// * The single consumer group visibility, which changes even on sealed extents due to DLQ merge/purge operations
// * The list of owning store IDs, which (in the future) will change with re-replication, archival activities, etc.`